### ISSUE TITLE
How to handle `yuv420p10le`?

### ISSUE THREAD

**Issue by riaqn** [NVIDIA/VideoProcessingFramework#274]
I have some videos of `yuv420p10le` pixel format given by ffmpeg. Should I just specify `nvc.PixelFormat.NV12` when creating decoder? In that case the downloaded picture seems totally off-color. I suppose this means the decoder doesn't handle 10bit LE very well. But when I look at the code https://github.com/NVIDIA/VideoProcessingFramework/blob/885296a1066128d36719d699f7c76f77e9c081ea/PyNvCodec/TC/src/NvDecoder.cpp#L275

it seems to be handling depth correctly. I suppose then it has to be about the bytes order - does Nvidia codec SDK only supports BE?

**Comment by rarzumanyan** [comment]
Hi @riaqn 

High bit depth is not currently supported by VPF.
It’s not a HW problem.

**Comment by riaqn** [comment]
@rarzumanyan Thank you for your quick reply. However, I'm reading the nvidia sdk document and couldn't find a field about byte order when creating decoders. Or do you mean packet can simply be converted to BE in software before passing to the decoder?

**Comment by rarzumanyan** [comment]
@riaqn 

IIRC both LE and BE shall be supported.

An easy way to check would be to decode your file with ffmpeg using NV HW-accelerated decoders (h264_cuvid or hevc_cuvid). Also, refer to Nvdec support matrix to check if your GPU supports 10 bit decoding on HW level.

Encoded packet contains information required to reconstruct the samples of video frame (motion vectors, transform coefficients and such), it has nothing to do with endiannes.

P. S.
Nvidia developers forum would be a better place to address this question.

**Comment by riaqn** [comment]
@rarzumanyan Thank you for the tip. I used `ffmpeg` to decode:
```
ffmpeg -c:v hevc_cuvid  -i 'test.mp4' -pix_fmt yuv420p10le out.yuv
```
(I can observe GPU usage high, therefore definitely using GPU)
then play the raw YUV by
```
ffplay -f rawvideo -pixel_format yuv420p10le -video_size 7260x3630 out.yuv
```
and the picture seems correct.

Therefore I guess both the GPU and the driver are 10bit HEVC ready. Do you think adding support to VPF will be a big workload? I'm not in urgent need(only a few videos of my media collection is 10bit hevc) so I can wait until you have time.

**Comment by rarzumanyan** [comment]
@riaqn 

>Do you think adding support to VPF will be a big workload?
>I'm not in urgent need

It's not that much of a job, but not the highest prio as well. Most of the effort would be to modify existing `Surface` classes to support higher bit depth.

I'll take this up once I refactor code or the bigger issue pops up which requires the high bit depth support.

**Comment by riaqn** [comment]
@rarzumanyan Sounds good! Looking forward to it.

**Comment by rarzumanyan** [comment]
Hi @Renzzauw  @riaqn 

Yes, it's still something I'd like to support ;)
Just under water with other tasks at the moment, my apologies.

**Comment by riaqn** [comment]
I don't mean to hijack the thread, but my library https://github.com/riaqn/python-nvidia-codec supports that. :smile:  but of course you have to used a patched pyav.

**Comment by Renzzauw** [comment]
@rarzumanyan Thanks for the update!

@riaqn Thanks for sharing that, it sounds like a useful library and it being written fully in python. Nice work! 
My current video processing application depends on VPF at the moment, and I am not in a rush to support other pixel formats not supported by VPF, but I'll keep an eye out on the developments of your library, it might be useful in the future!

**Comment by rarzumanyan** [comment]
@Renzzauw, @riaqn 

Please check out `issue_274` feature branch. It adds 10 and 12 bit decoding support.
I only checked 10 bit HEVC (don't have too many HBD videos nearby). This branch introduces 2 new pixel formats:

1. `P10`
2. `P12`

Both are basically 16-bit NV12. Please test it and LMK if it fulfills the request, I'll merge to `master` then.

**Comment by Renzzauw** [comment]
Hi @rarzumanyan, thanks for checking this out so quickly! I'm currently in the process of testing your changes. 

I'm testing it by adjusting my current video processing pipeline. I currently use the conversion chain NV12 -> YUV420 -> RGB -> RGB_PLANAR, how do I adjust this to work with P10? 

I tried the following conversions, but it gives me "Unsupported pixel format conversion errors":
- P10 -> NV12
- P10 -> YUV420

If I keep the chain as is, it runs problem-free, but the output video looks odd (which I would expect, given it's not 8-bit but a 10-bit video I'm inputting).

**Comment by rarzumanyan** [comment]
>how do I adjust this to work with P10?

By adding HBD color conversion support ;)
Existing color conversion part of VPF only supports 8 bit color.
Alternatively, some sort of 10 bit > 8 bit range mapping has to be done (doesn't exist yet in VPF yet).

**Comment by rarzumanyan** [comment]
@Renzzauw 

>Would it be possible for you to support such conversions or range mappings?

NPP does support such type of [conversion](https://docs.nvidia.com/cuda/npp/group__image__scale__to__lower__bit__depth.html#ga1cca684f7520c7f04aca57ec2d4c64d5) but there's a caveat here. 
Scaling factor is supported only for 32-bit floating point images:
```
dstPixelValue = dstMinRangeValue + scaleFactor * (srcPixelValue - srcMinRangeValue)
scaleFactor = (dstMaxRangeValue - dstMinRangeValue) / (srcMaxRangeValue - srcMinRangeValue).
```

E. g. this is the function for conversion from 16 to 8 bit:
```cpp
NppStatus nppiScale_16u8u_C1R_Ctx ( 
 const Npp16u *pSrc, 
 int nSrcStep, 
 Npp8u *pDst,
 int nDstStep,
 NppiSize oSizeROI,
 NppHintAlgorithm hint,
 NppStreamContext nppStreamCtx)  
```

WIll make the same type of conversion disregarding the actual bit depth (e. g. 10 or 12 bit).
I'm searching for a way to implement a proper mapping of 10- and 12-bit images to 8 bit but so far I didn't find anything.

**Comment by riaqn** [comment]
I don't mean to interrupt, but the way I did it was meta-programming a cuda program to do the conversion, support 16bit source:
https://github.com/riaqn/python-nvidia-codec/blob/main/src/nvidia_codec/color.py

**Comment by rarzumanyan** [comment]
@riaqn 
I like the idea but there are couple important things to keep in mind:

- FFMpeg is multi-license and mostly under LGPL
- Your project is under GPL3
- VPF is under Apache2

Have no idea how to copy-paste the code without running into legal issues ;)
I don't own this project and shall be extra careful regarding this matter.

Also, since it's a Video processing framework, it would be nice to pay some attention to bit depth conversion because this is part of HDR formats which may have different mapping algos between bit depths. Nvdec and Nvenc natively support HBD and hence may be used to decode, say HDR sequences. I would probably have to look onto more generic CUDA-accelerated conversion.

**Comment by riaqn** [comment]
@rarzumanyan Yes I understand your concern. 

WRT bit depth conversion - you mean I can't simply truncate 16bit and get the 8bit?

**Comment by riaqn** [comment]
@rarzumanyan yes by 'truncation' I mean bit truncation. I.e.: take the most significant 8 bits of the 10bits, which is the 'naive way' you mentioned. Could you maybe provide some examples of why this is bad? Thank you.

**Comment by rarzumanyan** [comment]
@Renzzauw 

With couple NPP tricks I've implemented simplest `p10>nv12` and `p12>nv12` cconv by integer division and range conversion.
To naked eye cconv looks good, so IMO it can be used for until someone comes up with better conversion algo.

**Comment by riaqn** [comment]
@Renzzauw I'm not sure if this is useful to you, but it seems that pytorch 0.12 now includes GPU decoding already: https://pytorch.org/vision/stable/generated/torchvision.io.VideoReader.html#torchvision.io.VideoReader

And even if you don't pytorch, there is also https://github.com/dmlc/decord#bridges-for-deep-learning-frameworks

But I don't really what are the catches of using these. Maybe @rarzumanyan  will know?

**Comment by Renzzauw** [comment]
@rarzumanyan 
Thanks for looking into implementing these conversions! I'll check out the new changes.

@riaqn 
Thanks for providing these resources, these are certainly helpful and welcome! 

I did some research into various python decoding & encoding options, and during that time VPF appeared most appealing to me, as it also provided encoding solutions, hence I probably didn't look that deep into torchvision's VideoReader, but I do wonder how these two compare performance-wise.

Decord also seems nice, but by the looks of the issue board, it is no longer actively being maintained?

Anyway, appreciate the help from both of you!

**Comment by riaqn** [comment]
@Renzzauw yes Decord seems to need significant patch to work with ffmpeg5.
and the decoder in torchvision is not very optimized. For example, they don't provide an interface to do batch screenshooting into a N*3*H*W tensor, which would be much more efficient than getting frames one by one.

**Comment by ADongGu** [comment]
> @Renzzauw, @riaqn
> 
> Please check out `issue_274` feature branch. It adds 10 and 12 bit decoding support. I only checked 10 bit HEVC (don't have too many HBD videos nearby). This branch introduces 2 new pixel formats:
> 
> 1. `P10`
> 2. `P12`
> 
> Both are basically 16-bit NV12. Please test it and LMK if it fulfills the request, I'll merge to `master` then.

me too , i want add decode yuv420p10le,   Here are my steps：
1： Adding the surface class and changing the update function and PlanePtr function should be OK, because only these two are different when decoding the output

**Comment by ADongGu** [comment]
> @Renzzauw,@riaqn
> 
> 请查看`issue_274`功能分支。它增加了 10 位和 12 位解码支持。 我只检查了 10 位 HEVC（附近没有太多 HBD 视频）。该分支引入了 2 种新的像素格式：
> 
> 1. `P10`
> 2. `P12`
> 
> 两者基本上都是16位NV12。请测试它和LMK是否满足请求，我将合并到`master`那时。

Boss, I have some questions after reading the code, that is, to see that you SurfaceP10 just inherited the SurfaceNV12, there are other details of the change, feel the most critical
1. virtual CUdeviceptr PlanePtr(uint32_t planeNumber = 0U) = 0;
2. virtual bool Update(SurfacePlane *pPlanes, size_t planesNum) = 0;
Nothing has changed.

**Comment by ADongGu** [comment]
> @Renzzauw
> 
> 通过几个 NPP 技巧，我通过整数除法和范围转换实现了 simple`p10>nv12`和`p12>nv12`cconv。 肉眼看来 cconv 看起来不错，所以 IMO 可以使用它，直到有人想出更好的转换算法。
> 
> @riaqn
> 
> > 您能否提供一些例子来说明为什么这很糟糕？
> 
> 这还不错，这只是最简单直接的方法。 根据我的经验（可能不相关，我没有广泛使用 HBD 视频），HBD 视频用于编辑等，并且用户通常更喜欢从 HBD 到 8 位进行更复杂的映射。LSB 截断作为增加 VPF 可以处理的视频量的一种措施还可以，但对于高级编辑来说可能不是最佳选择。

My cuda version is 11.3. error: 'nppiDivC_16u_C1RSfs_Ctx' was not declared in this scope. What is your version?

Also, so that 10 bits to 8 bits, the encoding is still 8-bit encoding, if you want to directly 10-bit encoding need to deal with.

Hope to get your reply, best wishes.

**Comment by ADongGu** [comment]
> @Renzzauw
> 
> With couple NPP tricks I've implemented simplest `p10>nv12` and `p12>nv12` cconv by integer division and range conversion. To naked eye cconv looks good, so IMO it can be used until someone comes up with better conversion algo.
> 
> @riaqn
> 
> > Could you maybe provide some examples of why this is bad?
> 
> It's not bad it's just the simplest straightforward approach. From my experience (which may not be relevant, I've not extensively worked with HBD videos) HBD videos are used for editing and such and often users prefer to do a more sophisticated mapping from HBD to 8 bit. LSB truncation is okay-ish as a measure to increase amount of videos VPF can handle but may be sub-optimal for advanced editing.

My cuda version is 11.3. error: 'nppiDivC_16u_C1RSfs_Ctx' was not declared in this scope. What is your version?

Also, so that 10 bits to 8 bits, the encoding is still 8-bit encoding, if you want to directly 10-bit encoding need to deal with.

Hope to get your reply, best wishes.

**Comment by ADongGu** [comment]
> @Renzzauw
> 
> With couple NPP tricks I've implemented simplest `p10>nv12` and `p12>nv12` cconv by integer division and range conversion. To naked eye cconv looks good, so IMO it can be used until someone comes up with better conversion algo.
> 
> @riaqn
> 
> > Could you maybe provide some examples of why this is bad?
> 
> It's not bad it's just the simplest straightforward approach. From my experience (which may not be relevant, I've not extensively worked with HBD videos) HBD videos are used for editing and such and often users prefer to do a more sophisticated mapping from HBD to 8 bit. LSB truncation is okay-ish as a measure to increase amount of videos VPF can handle but may be sub-optimal for advanced editing.

My cuda version is 11.3. error: 'nppiDivC_16u_C1RSfs_Ctx' was not declared in this scope. What is your version?

Also, if you go from 10 bits to 8 bits, the code will still be 8 bits.
How to deal with it if you want to directly encode 10 bits.

Hope to get your reply, best wishes.

**Comment by riaqn** [comment]
My library is here, feel free to try: https://github.com/riaqn/python-nvidia-codec
